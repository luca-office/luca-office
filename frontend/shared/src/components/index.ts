export * from "./accordion-card/accordion-card"
export * from "./app-header/app-header"
export * from "./app-notification/app-notification"
export * from "./auth-info-column/auth-info-column"
export * from "./auth-info-column/auth-info-column-container"
export * from "./automated-coding-item-rule-field/automated-coding-item-rule-field"
export * from "./binary-upload-input/binary-upload-input"
export * from "./binary-viewer"
export * from "./book-table-of-contents/book-table-of-contents"
export * from "./breadcrumb/breadcrumb"
export * from "./button/button"
export * from "./card"
export * from "./chat"
export * from "./checkbox/checkbox"
export * from "./columns/columns"
export * from "./content"
export * from "./content-image/content-image"
export * from "./content-loading-indicator/content-loading-indicator"
export * from "./content-video/content-video"
export * from "./create-main-dimension-modal/create-main-dimension-modal"
export * from "./date-picker"
export * from "./date-range-picker"
export * from "./delete-or-archive-entity-button/delete-or-archive-entity-button"
export * from "./desktop/desktop"
export * from "./desktop/desktop-background"
export * from "./desktop/desktop-background-overlay"
export * from "./detail-view-card/detail-view-card"
export * from "./detail-view-header/detail-view-header"
export * from "./error-boundary/error-boundary"
export * from "./event-questionnaire/event-questionnaire"
export * from "./feature-disabled-marker/feature-disabled-marker"
export * from "./file-drop-zone/file-drop-zone"
export * from "./file-explorer"
export * from "./files-and-directories-detail-empty/files-and-directories-detail-empty"
export * from "./files-and-directories-detail-view/files-and-directories-detail-view"
export * from "./finish-modal/finish-modal"
export * from "./form-error-text/form-error-text"
export * from "./form-field-label/form-field-label"
export * from "./header-carousel/header-carousel"
export * from "./header-carousel/header-carousel-container"
export * from "./header-content-container/header-content-container"
export * from "./header-footer-container/header-footer-container"
export * from "./icon/icon"
export * from "./image-viewer/image-viewer"
export * from "./input"
export * from "./label/label"
export * from "./loading-indicator/loading-indicator"
export * from "./local-file-chip/local-file-chip"
export * from "./markdown-editor/markdown-editor"
export * from "./markdown-renderer/markdown-renderer"
export * from "./markdown-text/markdown-text"
export * from "./media-file-preview/media-file-preview"
export * from "./modal/modal"
export * from "./module-progress-indicator/module-progress-indicator"
export * from "./module-start-overlay/module-start-overlay"
export * from "./office-window/office-window"
export * from "./orly/orly"
export * from "./orly-button/orly-button"
export * from "./overlay/overlay"
export * from "./paper/paper"
export * from "./participant-questionnaire-detail/participant-questionnaire-detail"
export * from "./participant-scenario-detail/participant-scenario-detail"
export * from "./participant-scenario-detail/participant-scenario-detail"
export * from "./participant-scenario-detail/questionnaire-results-overlay"
export * from "./pdf-viewer/pdf-view"
export * from "./pdf-viewer/pdf-viewer"
export * from "./progress-bar/progress-bar"
export * from "./progress-bar-column/progress-bar-column"
export * from "./progress-time-clock/progress-time-clock"
export * from "./questionnaire"
export * from "./radio-button/radio-button"
export * from "./rating"
export * from "./react-portal/portal-utils"
export * from "./react-portal/react-portal"
export * from "./readonly-action-field/readonly-action-field"
export * from "./reference-book-chapters"
export * from "./report-participant-overview-table/report-participant-overview-table"
export * from "./reporting-final-score"
export * from "./reporting-progress-info/reporting-progress-info"
export * from "./reporting-questionnaire-overlay/reporting-questionnaire-overlay"
export * from "./reporting-scenario-overlay/reporting-scenario-overlay"
export * from "./reset-password-link/reset-password-link"
export * from "./scenario-coding-model-table-of-contents/scenario-coding-model-table-of-contents"
export * from "./scenario-info-card/info-tag"
export * from "./scenario-info-card/scenario-info-card"
export * from "./scenario-info-card/scenario-info-card"
export * from "./select/custom-select"
export * from "./selectable-card/selectable-card"
export * from "./selection-in-preview-footer/selection-in-preview-footer"
export * from "./settings-footer-card/settings-footer-card"
export * from "./slide-menu/slide-menu"
export * from "./spreadsheet"
export * from "./spreadsheet-preview/spreadsheet-file-preview"
export * from "./sub-header/sub-header"
export * from "./tab-button-bar/tab-button-bar"
export * from "./table/table"
export * from "./table/table-container"
export * from "./table-of-content"
export * from "./table-of-content/table-of-contents-entry/table-of-contents-entry"
export * from "./table-of-content/table-of-contents-entry-text-placeholder/table-of-contents-entry-text-placeholder"
export * from "./tag/tag"
export * from "./tag-placeholder/tag-placeholder"
export * from "./taskbar/taskbar"
export * from "./text-document-preview/text-document-preview"
export * from "./textarea/textarea"
export * from "./tooltip/tooltip"
export * from "./typography/typography"
export * from "./update-text-document/update-text-document"
export * from "./upload-file-modal/upload-file-modal"
export * from "./video-viewer/video-viewer"
export * from "./viewer-tools"
export * from "./warning-tooltip/warning-tooltip"
export * from "./window-action-bar/window-action-bar"
