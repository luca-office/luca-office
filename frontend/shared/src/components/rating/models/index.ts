export * from "./criterion-metadata"
export * from "./rating-coding-criterion"
export * from "./scenario-criterion-selection"
export * from "./table-entity"
