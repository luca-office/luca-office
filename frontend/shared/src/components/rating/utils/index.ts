export * from "./coding-criterion"
export * from "./coding-dimension"
export * from "./coding-item"
export * from "./question"
export * from "./score"
export * from "./scoring-type"
