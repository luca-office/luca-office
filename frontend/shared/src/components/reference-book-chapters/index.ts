export * from "./article/reference-book-article"
export * from "./chapter/reference-book-chapter"
