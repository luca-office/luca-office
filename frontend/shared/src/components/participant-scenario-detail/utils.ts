import {FreetextAnswerFragment} from "../../graphql/generated/FreetextAnswerFragment"
import {QuestionType} from "../../graphql/generated/globalTypes"
import {Questionnaire, RuntimeSurveyResults} from "../../models"
import {ParticipantQuestionAnswer, QuestionnaireParticipantResults} from "../questionnaire"

export const runtimeSurveyResultsToQuestionnaireParticipantResults = (
  invitationId: UUID,
  questionnaire: Questionnaire,
  runtimeSurveyResults: RuntimeSurveyResults,
  freeTextAnswers: FreetextAnswerFragment[]
): QuestionnaireParticipantResults => {
  const answers = runtimeSurveyResults.questionResults.reduce<Record<UUID, ParticipantQuestionAnswer>>(
    (answersAccumulator, result) => {
      const question = questionnaire.questions.find(question => question.id === result.questionId)
      const participantResult = result.participantResults.find(results => results.invitationId === invitationId)
      const freeTextAnswer = freeTextAnswers.find(
        answer => answer.questionId === result.questionId && answer.surveyInvitationId === invitationId
      )?.text

      switch (question?.questionType) {
        case QuestionType.SingleChoice: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.SingleChoice,
              questionId: question.id,
              selectedAnswerId: participantResult?.selectedAnswerIds[0] ?? "",
              isFreetextAnswerSelected: participantResult?.wasFreetextAnswerSelected ?? false,
              freeTextAnswer: freeTextAnswer
            }
          }
        }

        case QuestionType.MultipleChoice: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.MultipleChoice,
              questionId: question.id,
              selectedAnswerIds: participantResult?.selectedAnswerIds ?? [],
              isFreetextAnswerSelected: participantResult?.wasFreetextAnswerSelected ?? false,
              freeTextAnswer: freeTextAnswer
            }
          }
        }

        case QuestionType.FreeText: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.FreeText,
              questionId: question.id,
              freeTextAnswer: freeTextAnswer ?? ""
            }
          }
        }

        default:
          return answersAccumulator
      }
    },
    {}
  )

  return {questionnaireId: runtimeSurveyResults.questionnaireId, answers}
}
