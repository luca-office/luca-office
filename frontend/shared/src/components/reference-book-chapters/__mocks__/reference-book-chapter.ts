import {ReferenceBookContentType} from "../../../enums"
import {MimeType} from "../../../graphql/generated/globalTypes"
import {ReferenceBookChapter} from "../../../models"

export const referenceBookChapterMock = (
  id = "2dbdae38-581a-430c-b635-c56b3ce4defa",
  publishedAt = ""
): ReferenceBookChapter => ({
  __typename: "ReferenceBookChapter",
  id: id,
  title: "Venture-Capital",
  createdAt: "2020-07-21T12:15:48.373Z",
  publishedAt: publishedAt,
  description:
    "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.",
  author: {
    __typename: "UserAccount",
    id: "2dbdae38-581a-430c-b635-c56b3ce4defa",
    firstName: "First",
    lastName: "Last"
  },
  articles: [
    {
      __typename: "ReferenceBookArticle",
      id: "fdspof-sdfpgofjds-sdf",
      title: "Begriff",
      position: 1,
      referenceBookChapterId: "2dbdae38-581a-430c-b635-c56b3ce4defa",
      contents: [
        {
          __typename: "ReferenceBookContent",
          contentType: ReferenceBookContentType.ImageContent,
          id: "sdgqwesdfpisdf",
          text: "sfdfdf",
          imageUrl:
            "https://images.unsplash.com/photo-1593642703055-4b72c180d9b5?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80",
          position: 2,
          createdAt: new Date(2020, 10, 5).toISOString(),
          modifiedAt: new Date(2020, 10, 15).toISOString(),
          imageBinaryFileId: "1c6d505e-06db-4ab2-ac0b-34f3b69e7626",
          videoBinaryFileId: null,
          referenceBookArticleId: "fdspof-sdfpgofjds-sdf",
          videoUrl: null,
          imageBinaryFile: {
            __typename: "BinaryFile",
            id: "1c6d505e-06db-4ab2-ac0b-34f3b69e7626",
            createdAt: new Date(2020, 10, 5).toISOString(),
            modifiedAt: new Date(2020, 10, 15).toISOString(),
            filename: "photo-1593642703055-4b72c180d9b5",
            fileSize: 1,
            mimeType: MimeType.ImagePng,
            url:
              "https://images.unsplash.com/photo-1593642703055-4b72c180d9b5?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80"
          },
          videoBinaryFile: null,
          pdfBinaryFileId: null,
          pdfBinaryFile: null
        }
      ]
    },
    {
      __typename: "ReferenceBookArticle",
      id: "fdspof-sdfpgofjds-sdf-sdfsd",
      title: "ertragspotenziale und Risiken von Venture-Capital-Finanzierungen",
      position: 2,
      referenceBookChapterId: "2dbdae38-581a-430c-b635-c56b3ce4defa",
      contents: [
        {
          __typename: "ReferenceBookContent",
          contentType: ReferenceBookContentType.ImageContent,
          id: "sdgqwesdfpisdf",
          text: "sfdfdf",
          imageUrl:
            "https://images.unsplash.com/photo-1593642703055-4b72c180d9b5?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80",
          position: 2,
          createdAt: new Date(2020, 10, 5).toISOString(),
          modifiedAt: new Date(2020, 10, 15).toISOString(),
          imageBinaryFileId: "1c6d505e-06db-4ab2-ac0b-34f3b69e7626",
          videoBinaryFileId: null,
          referenceBookArticleId: "fdspof-sdfpgofjds-sdf",
          videoUrl: null,
          imageBinaryFile: {
            __typename: "BinaryFile",
            id: "1c6d505e-06db-4ab2-ac0b-34f3b69e7626",
            createdAt: new Date(2020, 10, 5).toISOString(),
            modifiedAt: new Date(2020, 10, 15).toISOString(),
            filename: "photo-1593642703055-4b72c180d9b5",
            fileSize: 1,
            mimeType: MimeType.ImagePng,
            url:
              "https://images.unsplash.com/photo-1593642703055-4b72c180d9b5?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80"
          },
          videoBinaryFile: null,
          pdfBinaryFileId: null,
          pdfBinaryFile: null
        },
        {
          __typename: "ReferenceBookContent",
          contentType: ReferenceBookContentType.ImageContent,
          id: "sdgqwesdfpisdf",
          text: "sfdfdf",
          imageUrl:
            "https://images.unsplash.com/photo-1593642703055-4b72c180d9b5?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80",
          position: 2,
          createdAt: new Date(2020, 10, 5).toISOString(),
          modifiedAt: new Date(2020, 10, 15).toISOString(),
          imageBinaryFileId: "1c6d505e-06db-4ab2-ac0b-34f3b69e7626",
          videoBinaryFileId: null,
          referenceBookArticleId: "fdspof-sdfpgofjds-sdf",
          videoUrl: null,
          imageBinaryFile: {
            __typename: "BinaryFile",
            id: "1c6d505e-06db-4ab2-ac0b-34f3b69e7626",
            createdAt: new Date(2020, 10, 5).toISOString(),
            modifiedAt: new Date(2020, 10, 15).toISOString(),
            filename: "photo-1593642703055-4b72c180d9b5",
            fileSize: 1,
            mimeType: MimeType.ImagePng,
            url:
              "https://images.unsplash.com/photo-1593642703055-4b72c180d9b5?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80"
          },
          videoBinaryFile: null,
          pdfBinaryFileId: null,
          pdfBinaryFile: null
        }
      ]
    }
  ]
})
