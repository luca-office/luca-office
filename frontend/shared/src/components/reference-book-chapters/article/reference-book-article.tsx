import {css} from "@emotion/react"
import * as React from "react"
import {
  Card,
  CardContent,
  CardFooter,
  CardHeader,
  ContentImage,
  ContentVideo,
  Icon,
  LoadingIndicator,
  MarkdownText,
  Text
} from "../../../components"
import {IconName, ReferenceBookContentType} from "../../../enums"
import {ReferenceBookArticle as ReferenceBookArticleModel, ReferenceBookArticleContent} from "../../../models"
import {AutomatedCriterionReferenceBooksConfig} from "../../../office-tools"
import {
  backgroundColor,
  buttonHeight,
  CustomStyle,
  spacing,
  spacingLarge,
  spacingMedium,
  spacingSmall,
  TextSize,
  tocFooterHeight
} from "../../../styles"
import {useLucaTranslation} from "../../../translations"
import {Option} from "../../../utils"
import {ContentPdf} from "../../content-pdf/content-pdf"
import {SelectionInPreviewFooter} from "../../selection-in-preview-footer/selection-in-preview-footer"

export interface ReferenceBookArticleProps extends CustomStyle {
  readonly article: Option<ReferenceBookArticleModel>
  readonly chapterTitle?: string
  readonly dataLoading: boolean
  readonly selectImage?: (id: UUID, url: string, title?: string) => void
  readonly selectVideo?: (id: UUID, url: string, title?: string) => void
  readonly openPdf?: (id: UUID, url: string, title?: string) => void
  readonly automatedCriterionConfig?: AutomatedCriterionReferenceBooksConfig
}

export const ReferenceBookArticle: React.FunctionComponent<ReferenceBookArticleProps> = ({
  article: articleOption,
  chapterTitle,
  customStyles,
  dataLoading,
  selectImage,
  selectVideo,
  openPdf,
  automatedCriterionConfig
}) => {
  const {t} = useLucaTranslation()
  const renderContent = (
    article: ReferenceBookArticleModel,
    content: ReferenceBookArticleContent,
    contentIndex: number
  ) => {
    switch (content.contentType) {
      case ReferenceBookContentType.ImageContent:
        return content.imageUrl ? (
          <ContentImage
            key={content.id}
            imageUrl={content.imageUrl}
            onClick={() =>
              content.imageBinaryFileId !== null &&
              content.imageUrl !== null &&
              selectImage?.(content.imageBinaryFileId, content.imageUrl, `${contentIndex + 1} - ${article.title}`)
            }
          />
        ) : null
      case ReferenceBookContentType.VideoContent:
        return content.videoUrl ? (
          <ContentVideo
            key={content.id}
            videoUrl={content.videoUrl}
            onClick={() =>
              content.videoBinaryFileId !== null &&
              content.videoUrl !== null &&
              selectVideo?.(content.videoBinaryFileId, content.videoUrl, `${contentIndex + 1} - ${article.title}`)
            }
          />
        ) : null
      case ReferenceBookContentType.PdfContent:
        return content?.pdfBinaryFile ? (
          <ContentPdf
            key={content.id}
            openPdf={pdf => pdf.forEach(pdf => openPdf?.(pdf.id, pdf.url, pdf.filename))}
            title={content.pdfBinaryFile.filename}
            pdfFile={Option.of(content.pdfBinaryFile)}
          />
        ) : null
      case ReferenceBookContentType.TextContent:
        return content.text ? <MarkdownText key={content.id} markdown={content.text} /> : null
      default:
        return null
    }
  }
  return (
    <Card customStyles={[styles.card(automatedCriterionConfig !== undefined), customStyles]}>
      <CardHeader hasShadow hasGreyBackground>
        <Icon name={IconName.Book} hasSpacing />
        {articleOption.map(article => article.title).orNull()}
        <Text customStyles={styles.chapterTitle} size={TextSize.Medium}>
          {chapterTitle}
        </Text>
      </CardHeader>
      <CardContent customStyles={styles.scrollable}>
        <div css={styles.content(!selectImage)}>
          <Text customStyles={styles.title}>{articleOption.map(article => article.title).orNull()}</Text>
          {dataLoading ? (
            <LoadingIndicator customStyles={styles.loader} />
          ) : (
            articleOption
              .map(article => (
                <div>{article.contents.map((content, index) => renderContent(article, content, index))}</div>
              ))
              .orNull()
          )}
        </div>
      </CardContent>
      {automatedCriterionConfig !== undefined &&
        articleOption
          .map(article => (
            <SelectionInPreviewFooter
              textKey="coding_models__automated_item_document_view_reference_books_selection_footer"
              headingKey="coding_models__automated_item_document_view_column_header"
              icon={IconName.Book}
              customLabelStyles={styles.selectionFooter}
              onConfirm={() => automatedCriterionConfig?.onArticleClick?.(article.id)}
              title={t("reference_book__title_article")}
            />
          ))
          .orNull()}

      <CardFooter customStyles={automatedCriterionConfig !== undefined ? styles.footer : undefined} />
    </Card>
  )
}

const styles = {
  title: css({
    fontSize: TextSize.Larger
  }),
  footer: css({
    height: "auto"
  }),
  card: (hasSelectionFooter: boolean) =>
    css({
      borderColor: backgroundColor,
      overflow: "hidden",
      height: hasSelectionFooter ? undefined : `calc(100vh - ${buttonHeight}px - ${tocFooterHeight}px)`
    }),
  content: (readonly: boolean) =>
    css({
      padding: spacing(spacingMedium, spacingMedium, spacingMedium, spacingLarge),

      ".image": {
        cursor: readonly ? "default" : "pointer"
      }
    }),
  chapterTitle: css({
    opacity: 0.5,
    marginLeft: "auto"
  }),
  loader: css({
    margin: `${spacingLarge}px auto`
  }),
  scrollable: css({
    overflowY: "auto"
  }),
  selectionFooter: css({
    marginLeft: spacingSmall
  })
}
