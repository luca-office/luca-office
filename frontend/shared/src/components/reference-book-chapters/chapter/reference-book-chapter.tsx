import * as React from "react"
import {
  Card,
  CardContent,
  CardFooter,
  CardHeader,
  ColumnProps,
  Heading,
  Icon,
  LoadingIndicator,
  TableContainer,
  Text
} from "../../../components"
import {HeadingLevel, IconName} from "../../../enums"
import {ReferenceBookArticle, ReferenceBookChapter as ReferenceBookChapterModel} from "../../../models"
import {CustomStyle, Flex, FontWeight, TextSize} from "../../../styles"
import {useLucaTranslation} from "../../../translations"
import {Option, sortByPosition} from "../../../utils"
import {chapterStyles} from "./reference-book-chapter-style"

export interface ReferenceBookChapterProps extends CustomStyle {
  readonly chapter: Option<ReferenceBookChapterModel>
  readonly dataLoading: boolean
  readonly selectArticle: (id: Option<UUID>) => void
  readonly isBundled?: boolean
  readonly headerInfo?: string
  readonly headerChild?: React.ReactNode
}

export const ReferenceBookChapter: React.FunctionComponent<ReferenceBookChapterProps> = ({
  chapter: chapterOption,
  customStyles,
  dataLoading,
  headerChild,
  headerInfo,
  isBundled = false,
  selectArticle
}) => {
  const {t} = useLucaTranslation()

  const articleColumn: ColumnProps<ReferenceBookArticle> = {
    key: "data",
    content: (article: ReferenceBookArticle) => (
      <div css={Flex.row}>
        <Icon name={IconName.BookPage} hasSpacing /> {article.title}
      </div>
    ),
    header: (
      <div css={Flex.row}>
        <Icon name={IconName.BookPage} hasSpacing /> {t("reference_book__chapter_contents_column")}
      </div>
    )
  }
  const handleTableClick = (entity: ReferenceBookArticle) => selectArticle(Option.of(entity.id))

  return (
    <Card customStyles={[chapterStyles.card, customStyles]}>
      <CardHeader customStyles={chapterStyles.header} hasShadow hasGreyBackground>
        <div css={chapterStyles.row}>
          <Icon name={IconName.Book} hasSpacing />
          {chapterOption.map(chapter => chapter.title).getOrElse(t("reference_book__title_placeholder"))}
        </div>
        <div css={chapterStyles.row}>
          {headerInfo && (
            <Text size={TextSize.Medium} customStyles={chapterStyles.headerInfo}>
              {headerInfo}
            </Text>
          )}
          {headerChild}
        </div>
      </CardHeader>
      <CardContent customStyles={chapterStyles.scrollable}>
        {dataLoading ? (
          <LoadingIndicator customStyles={chapterStyles.loader} />
        ) : (
          chapterOption
            .map(chapter => (
              <div css={chapterStyles.content}>
                <Heading level={HeadingLevel.h1} customStyles={chapterStyles.headline}>
                  {chapter.title}
                </Heading>
                <Text size={TextSize.Medium} customStyles={chapterStyles.text(isBundled)}>
                  {chapter.description}
                </Text>
                {chapter.articles && chapter.articles.length > 0 && (
                  <React.Fragment>
                    <Heading
                      level={HeadingLevel.h3}
                      fontWeight={FontWeight.Bold}
                      customStyles={chapterStyles.label(isBundled)}>
                      {t("reference_book__chapter_contents_count", {count: chapter.articles.length})}
                    </Heading>
                    <TableContainer<ReferenceBookArticle>
                      customStyles={chapterStyles.table}
                      isBundled={isBundled}
                      entities={sortByPosition(chapter.articles)}
                      entityKey={(article: ReferenceBookArticle) => article.id}
                      columns={[articleColumn]}
                      onClick={handleTableClick}
                      placeHolderText={t("reference_books__table_of_contents_placeholder")}
                    />
                  </React.Fragment>
                )}
              </div>
            ))
            .orElse(
              Option.of(
                <div css={chapterStyles.placeholderContainer}>
                  <div>{t("reference_book__title_placeholder")}</div>
                  <div css={chapterStyles.hint}>{t("reference_book__content_placeholder")}</div>
                </div>
              )
            )
            .orNull()
        )}
      </CardContent>
      <CardFooter />
    </Card>
  )
}
