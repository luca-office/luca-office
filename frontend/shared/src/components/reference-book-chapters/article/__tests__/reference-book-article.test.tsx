// importing from direct file because of issues of babel loader and spyOn
import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {Card, CardContent, CardFooter, CardHeader, ContentImage, LoadingIndicator, Text} from "../../../../components"
import {ReferenceBookContentType} from "../../../../enums"
import {ReferenceBookArticle as ReferenceBookArticleModel, ReferenceBookChapter} from "../../../../models"
import {Option} from "../../../../utils"
import {referenceBookChapterMock} from "../../__mocks__/reference-book-chapter"
import {ReferenceBookArticle, ReferenceBookArticleProps} from "../reference-book-article"

const chapter: ReferenceBookChapter = referenceBookChapterMock()
const selectSpy = jest.fn()

const defaultProps: ReferenceBookArticleProps = {
  dataLoading: false,
  selectImage: selectSpy,
  chapterTitle: "my test chapter",
  article: Option.of(chapter.articles[0])
}

const getComponent = (props?: Partial<ReferenceBookArticleProps>) => (
  <ReferenceBookArticle {...defaultProps} {...props} />
)

describe("reference-book-article", () => {
  it("renders correctly", async () => {
    const component = create(getComponent())
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure", async () => {
    const component = shallow(getComponent())

    expect(component.find(Card)).toHaveLength(1)
    expect(component.find(CardHeader)).toHaveLength(1)
    expect(component.find(CardFooter)).toHaveLength(1)
    expect(component.find(LoadingIndicator)).toHaveLength(0)
    expect(component.find(CardContent).find(ContentImage)).toHaveLength(
      defaultProps.article
        .map(
          (article: ReferenceBookArticleModel) =>
            article.contents.filter(content => content.contentType === ReferenceBookContentType.ImageContent).length
        )
        .getOrElse(0)
    )
    expect(component.find(CardContent).find(Text)).toHaveLength(
      defaultProps.article
        .map(
          (article: ReferenceBookArticleModel) =>
            article.contents.filter(content => content.contentType === ReferenceBookContentType.TextContent).length
        )
        // By default title is always there
        .getOrElse(0) + 1
    )
  })
  it("has correct structure (loading)", async () => {
    const component = shallow(getComponent({dataLoading: true}))

    expect(component.find(Card)).toHaveLength(1)
    expect(component.find(CardHeader)).toHaveLength(1)
    expect(component.find(CardFooter)).toHaveLength(1)
    expect(component.find(LoadingIndicator)).toHaveLength(1)
    expect(component.find(Text)).toHaveLength(2)
  })
})
