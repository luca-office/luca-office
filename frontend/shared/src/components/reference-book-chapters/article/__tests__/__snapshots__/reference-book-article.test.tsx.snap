// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reference-book-article renders correctly 1`] = `
<div
  className="css-1wpi4x0-Card"
>
  <div
    className="card-header css-7banf6-CardHeader"
  >
    <div
      className="css-cj5xss-Icon"
      dangerouslySetInnerHTML={
        {
          "__html": {},
        }
      }
    />
    Begriff
    <div
      className="css-pysl8o-Text"
    >
      my test chapter
    </div>
  </div>
  <div
    className="card-content css-fnpw2k-CardContent"
  >
    <div
      className="css-f23wob-ReferenceBookArticle"
    >
      <div
        className="css-16sst13-Text"
      >
        Begriff
      </div>
      <div>
        <div
          className="css-2rq2kz-ContentImage"
          onClick={[Function]}
        >
          <div
            className="image css-1pdo4of-ContentImage"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="card-footer css-gmu9kx-CardFooter"
  />
</div>
`;
